bigtable/bttest/grpc_interceptor.go

```diff
@@ -0,0 +1,294 @@
+			if strings.HasSuffix(info.FullMethod, gt.methodSuffix) && eRand <= gt.errorRate {
+				log.Printf("Injecting Emulator Error for Error Target %s", gt.String())
+				return status.Error(gt.grpcErrorCode, "Injected Emulator Error")
+			}
+		}
+
+		return nil
+	}
+}
+
+/*
+	latencyTargets define how long it will take for a method
+	to return at different percentiles. e.g. ReadRows:p50:100ms
+*/
+
+type latencyTargets []latencyTarget
+
+// Set() used by Flags lib to create new Latency Target
+func (lts *latencyTargets) Set(s string) error {
+	lt, err := newLatencyTarget(s)
+	if err != nil {
+		return err
+	}
+	*lts = append(*lts, *lt)
+	return nil
+}
+
+func (lts *latencyTargets) String() string {
+	var s []string
+	for _, v := range *lts {
+		s = append(s, v.String())
+	}
+	return strings.Join(s, ", ")
+}
+
+type latencyTarget struct {
+	methodSuffix     string
+	percentile       int32
+	expectedDuration time.Duration
+}
+
+// Create new latencyTarget from string like "ReadRows:p50:100ms"
+func newLatencyTarget(s string) (*latencyTarget, error) {
+	var lt latencyTarget
+	var err error
+
+	pieces := strings.Split(s, ":")
+	if len(pieces) != 3 {
+		return nil, fmt.Errorf("Expected Latency Target in form of: <method>:<percentile>:<duration>")
+	}
+	err = lt.setMethodSuffix(pieces[0])
+	if err != nil {
+		return nil, err
+	}
+	err = lt.setPercentile(pieces[1])
+	if err != nil {
+		return nil, err
+	}
+	if err = lt.setExpectedDuration(pieces[2]); err != nil {
+		return nil, err
+	}
+
+	return &lt, nil
+}
+
+func (lt *latencyTarget) setMethodSuffix(s string) error {
+	if isValidStreamMethodSuffix(s) {
+		lt.methodSuffix = s
+		return nil
+	}
+	return fmt.Errorf("Invalid method for Latency Target. Expected one of: %s", validStreamMethodSuffixes)
+}
+
+func isValidStreamMethodSuffix(s string) bool {
+	for _, v := range validStreamMethodSuffixes {
+		if s == v {
+			return true
+		}
+	}
+	return false
+}
+
+func (lt *latencyTarget) setPercentile(s string) error {
+	p := strings.TrimPrefix(s, "p")
+	i, err := strconv.Atoi(p)
+	if err != nil || (i < 0 || i > 99) {
+		return fmt.Errorf("Invalid latency percentile: %s. Expected integer between 0 and 99", s)
+	}
+	lt.percentile = int32(i)
+	return nil
+}
+
+func (lt *latencyTarget) setExpectedDuration(s string) error {
+	d, err := time.ParseDuration(s)
+	if err != nil {
+		return fmt.Errorf("Invalid latency duration: %s.\n%s", s, err)
+	}
+	lt.expectedDuration = d
+	return nil
+}
+
+func (lt *latencyTarget) String() string {
+	return fmt.Sprintf("%s:p%d:%s", lt.methodSuffix, lt.percentile, lt.expectedDuration)
+}
+
+/*
+	grpcErrorCodeTargets define how often each method should throw a given GRPC Error Code
+	e.g. "ReadRows:10%:14" will have a 10% chance to throw Error Code 14 (Unavailable) on each ReadRows request
+*/
+
+type grpcErrorCodeTargets []grpcErrorCodeTarget
+
+// Set() used by Flags lib to create new grpc error code target
+func (ets *grpcErrorCodeTargets) Set(s string) error {
```

Comment:
Add?

Reply:
+1 that Add() is a better name.
Used `Set()` to satisfy the ["Value" interface](https://golang.org/src/flag/flag.go?s=7440:7500#L281) from the flags package